neo4j-to-neptune/src/main/java/com/amazonaws/services/neptune/util/NeptuneBulkLoader.java

```diff
@@ -270,66 +273,117 @@ private CompletableFuture<Boolean> uploadFilesSequentially(File[] files, String
-                    System.err.println("Error message: " + throwable.getMessage());
-
-                    if (throwable.getCause() instanceof S3Exception) {
-                        S3Exception s3Exception = (S3Exception) throwable.getCause();
-                        System.err.println("S3 error code: " + s3Exception.awsErrorDetails().errorCode());
-                        System.err.println("S3 error message: " + s3Exception.awsErrorDetails().errorMessage());
-                        System.err.println("S3 status code: " + s3Exception.statusCode());
-                    }
-                    return false;
-                } else {
-                    System.err.println("Successfully uploaded " + file.getName() + " using Transfer Manager - ETag: " +
-                            completedUpload.response().eTag());
+            System.err.println("Initiating Transfer Manager upload with compression...");
+            Upload upload = transferManager.upload(uploadRequest);
+
+            // Wait for BOTH upload and compression to complete - fail if either fails
+            return CompletableFuture.allOf(upload.completionFuture(), compressionFuture)
+                .thenApply(ignored -> {
+                    System.err.println(
+                        "Successfully uploaded " + localFile.getName() +
+                        " (compressed) - ETag: " + upload.completionFuture().join().response().eTag());
                     return true;
+                })
+                .exceptionally(throwable -> {
+                    logUploadError(localFilePath, throwable);
+                    // Re-throw to maintain fail-fast behavior
+                    if (throwable instanceof RuntimeException) {
+                        throw (RuntimeException) throwable;
+                    } else {
+                        throw new RuntimeException("Upload or compression failed", throwable);
+                    }
+                })
+                .whenComplete((result, throwable) -> {
+                    // Always shutdown the executor service
+                    streamExecutor.shutdown();
+                });
+        }
+    }
+
+    /**
+     * Start the compression task in a background thread
+     */
+    protected CompletableFuture<Void> startCompressionTask(File localFile, PipedOutputStream pipedOut) {
+        return CompletableFuture.runAsync(() -> {
+            try (GZIPOutputStream gzipOut = new GZIPOutputStream(pipedOut);
+                 FileInputStream fis = new FileInputStream(localFile);
+                 BufferedInputStream fileIn = new BufferedInputStream(fis)) {
+
+                // Use same buffer size as GzipCompressUtils for consistency
+                byte[] buffer = new byte[8192];
+                int bytesRead;
+                while ((bytesRead = fileIn.read(buffer)) != -1) {
+                    gzipOut.write(buffer, 0, bytesRead);
                 }
```

Comment:
Would this be simpler?
```suggestion
fileIn.transferTo(gzipOut);
```

Reply:
good question, ill try it out

Reply:
works great actually!